### ISSUE TITLE
ARCH-1 — Radial bases, cutoffs, node embedding, and native spherical harmonics (reference-only ops)

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1559]
**Depends on:** INF-1 (#1550) · **Blocks:** ARCH-2 (#1561)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** The pure-math building blocks of the forward pass port nearly verbatim, and P0-7 already wrote their behavioural spec as tests. This ticket moves them into the v1 stack: the radial bases and cutoffs from `mace/modules/radial.py` (`BesselBasis:18`, `ChebychevBasis:61`, `GaussianBasis:89`, `PolynomialCutoff:113`, `ZBLBasis:149`, `AgnesiTransform:225`, `SoftTransform:285`, `RadialMLP:361`), the embedding blocks from `mace/modules/blocks.py` (`LinearNodeEmbeddingBlock:44`, `RadialEmbeddingBlock:394`), and the spherical harmonics, which legacy delegates to e3nn (`o3.SphericalHarmonics(sh_irreps, normalize=True, normalization="component")`, `mace/modules/models.py:163`, repeated at `:691`, `:952`, `:1257` and `mace/modules/extensions.py:1544`). e3nn is removed entirely in v1, so the spherical harmonics must be reimplemented as a **native plain-torch reference-only op**, written for clarity rather than speed. The frozen legacy `mace/` is read in place as the behavioural oracle; reproduce its numbers, never its structure.

**Interface & constraints:**

1. **Where things land** (all under `packages/mace-torch/src/mace_torch/`):
   - `nn/radial.py` — `BesselBasis`, `ChebychevBasis`, `GaussianBasis`, `PolynomialCutoff`, `ZBLBasis`, `AgnesiTransform`, `SoftTransform`, `RadialMLP`, `ChebyshevBasisGeneral`.
   - `nn/embedding.py` — `LinearNodeEmbeddingBlock`, `RadialEmbeddingBlock`.
   - `backends/reference/sph_harm.py` — the native spherical harmonics + reference radial basis (non-hot).
   No `@compile_mode('script')` on any block (legacy carries it on every one — `mace/modules/blocks.py:43,63,85,121,161,183,231,260,362,393,438,515,637,…`); no e3nn anywhere; `packages/` never imports the legacy `mace/` (parity tests under `tests/parity/` may).

2. **Reference-only op status.** `SphericalHarmonicsOp` and `RadialBasisOp` are *reference-only* in the backend contract: closed-form, cheap, never dispatched today. A backend *may* override them but is not required to (the factory may decline). The authoritative implementations are the ones this ticket writes; the torch-free spec (descriptors, tolerances) lives in `mace_core` and is defined by BKD-1 (#1560) — coordinate on descriptor field names, but do not block on BKD-1 (#1560): BKD-1 (#1560)'s reference backend wraps these implementations in its `make_spherical_harmonics` / `make_radial_basis` factories.

3. **Numeric semantics to preserve exactly** (P0-7's radial tests are the parity spec):
   - Every basis×cutoff product has value **and** first derivative → 0 at `r_max`, and is exactly zero beyond.
   - `RadialEmbeddingBlock` forward ordering: cutoff is computed on the **raw** edge lengths; the distance transform (Agnesi/Soft, when configured) is applied to the lengths **before** the basis; with `apply_cutoff=False` the block returns `(radial, cutoff)` unmultiplied, otherwise `(radial * cutoff, None)`. Replace the legacy `hasattr(self, "distance_transform")` dynamic checks with explicit configuration, but keep the numerics identical.
   - `ZBLBasis` matches the published ZBL formula (P0-7 asserts it for two element pairs with hand-computed values). Its placement relative to scale/shift is *not* this ticket's business — that asymmetry is ARCH-3 (#1562)'s `zbl_in_scale_shift`.
   - `LinearNodeEmbeddingBlock` maps one-hot element attributes to channel features; the attributes are scalar (l=0), so this is a plain weight matrix in torch — no equivariant machinery needed.
   - `ChebyshevBasisGeneral` (`mace/modules/extensions.py:1380`) is **not** a `--radial_type` variant and never was: it is the *magnetic-moment* radial basis, constructed with `r_max=0.0` and `include_constant=False` at `mace/modules/extensions.py:1553`. It ports here because it is pure radial math, but its consumer is the magnetic family (ARCH-5 (#1633)/MAG-1 (#1584)), and its `include_constant` switch is part of the contract.

4. **Spherical harmonics contract.** Real spherical harmonics up to the model `lmax`, matching e3nn's legacy convention: input vectors normalized (`normalize=True`), `"component"` normalization, and the e3nn output ordering the rest of the legacy model assumes. Differentiable to **second order** (force/stress training runs `grad(grad(E))` through the spherical harmonics and the radial embedding), verified by `gradgradcheck`.

5. **sphericart is already in-tree, and already load-bearing.** `sphericart` is the concrete candidate for an *accelerated* `SphericalHarmonicsOp` override: fast, CPU+GPU, differentiable to second order, and pinned by the `[magnetic]` extra as `sphericart-torch==1.0.9` (`setup.cfg`, #1244). Develop does not merely *depend* on it — the magnetic family **uses** it: `SHModule` (`mace/modules/extensions.py:1357`) wraps `sphericart.torch.SolidHarmonics(l_max)` and is instantiated at `:1562`, with a fixed `[2, 0, 1]` index permutation (xyz → z,x,y) as the recorded convention mapping to e3nn ordering. So the evaluation this ticket writes is not speculative: it decides whether the *same* op serves the edge harmonics too. The native plain-torch spherical harmonics stays the reference regardless of the outcome.

**Task:**
1. Port the eight radial classes to `mace_torch/nn/radial.py` in plain torch, P0-7 tests as the spec; add `ChebyshevBasisGeneral` beside them; port `LinearNodeEmbeddingBlock` and `RadialEmbeddingBlock` to `mace_torch/nn/embedding.py` preserving the forward semantics above.
2. Implement the native spherical harmonics in `mace_torch/backends/reference/sph_harm.py`: closed-form real spherical harmonics up to `lmax`, e3nn `normalize=True` / `"component"` conventions, double-backward correct.
3. Bring the P0-7 radial test cases over to `packages/mace-torch/tests/` (run against the ported modules) and add parity tests vs the frozen legacy `mace/` under `tests/parity/` for the spherical harmonics and `RadialEmbeddingBlock` (including `apply_cutoff=False` and the Agnesi/Soft transforms).
4. Add `gradcheck`/`gradgradcheck` coverage at fp64 for the spherical harmonics and each basis/cutoff w.r.t. the edge vectors/lengths.
5. Evaluate `sphericart`: fp64 parity vs the native implementation (values, first and second derivatives, convention mapping including the `[2, 0, 1]` permutation and the solid-vs-spherical distinction develop uses), CPU timing, and a short written recommendation on adopting it as the optional accelerated override. **GPU half measured** (A100, `lmax=3`, 200k edges, the pinned `sphericart-torch==1.0.9`) — parity against the *native* implementation is open — this ticket writes that implementation:
   - **Second derivatives are opt-in and OFF by default.** The installed signature is `SolidHarmonics(l_max, backward_second_derivatives=False)` / `SphericalHarmonics(l_max, backward_second_derivatives=False)`, and with the default `gradgradcheck` **fails** (Jacobian mismatch) while e3nn passes. MACE trains forces via `grad(grad(E))`, so a defaults drop-in would be silently wrong. With `backward_second_derivatives=True`, `gradgradcheck` passes.
   - **The flag costs 45 %**: fwd+bwd 0.640 → 0.931 ms (fp64). So the honest speedup is the one measured *with* it: **1.61× fwd+bwd** vs e3nn (1.496 ms), not the 2.42× the default suggests. Forward alone: 2.0× (fp64) / 3.6× (fp32).
   - **Convention mapping is a single scalar.** The e3nn/sphericart l-block norm ratio is 3.544908 = √(4π), constant across l=0..3 to 6e-15 — so `"component"` differs from sphericart's normalization by one global factor, not a per-l table.
   - **Packaging caveat:** the pinned 1.0.9 has no ABI-compatible prebuilt for torch 2.13 and must be built from source (~2.5 min); the built extension then needs a CUDA runtime on the compute node (`module load cuda`), not just torch's bundled one.
   - **TODO (open, needs the maintainer's call):** `SHModule` constructs `SolidHarmonics(l_max)` with the default, i.e. **second derivatives off**, and its output feeds `magmom_node_attrs = self.mag_solid_harmoics(data["magmom"])` (`mace/modules/extensions.py:1834`) on a tensor that `MagneticScaleShiftMACE.forward` marks `requires_grad_(True)` at `:1763`. Magforce *training* differentiates through that backward (`create_graph=training`), so develop's magnetic second-derivative path runs on the configuration sphericart documents as incorrect. This is a live defect, not a fixed one, so rule 3 does not settle it: v1 either reproduces it (bit-parity with develop, knowingly wrong gradients) or sets `backward_second_derivatives=True` (a declared numeric deviation). ARCH-5 (#1633) carries the decision; ARCH-1 (#1559) must surface the flag either way.

**Out of scope:** the interaction block, symmetric contraction, and `BaseMACE` assembly (ARCH-2 (#1561)); the `KernelBackend` Protocol, registry, and dispatched hot ops (BKD-1 (#1560)); actually wiring `sphericart` in as a shipped override (follow-up decided by the evaluation); the magnetic consumers of `ChebyshevBasisGeneral`/`SHModule` (ARCH-5 (#1633), MAG-1 (#1584)); any change to legacy `mace/`.

**Acceptance criteria:**
- [ ] The P0-7 radial test cases pass unchanged in expectation against the ported modules (cutoff smoothness, exact zero beyond `r_max`, ZBL formula, transform values + monotonicity, batched shape/dtype contracts).
- [ ] Native spherical harmonics match the frozen legacy e3nn `o3.SphericalHarmonics` on random unit and non-unit vectors up to the model `lmax` at fp64 tolerance.
- [ ] `gradcheck` **and** `gradgradcheck` pass at fp64 for the spherical harmonics and for each radial basis × cutoff composition.
- [ ] `RadialEmbeddingBlock` parity vs legacy on both return paths (`apply_cutoff` True/False) and with Agnesi/Soft transforms enabled.
- [ ] `ChebyshevBasisGeneral` matches legacy for both `include_constant` values, at the `r_max=0.0` configuration the magnetic family actually constructs.
- [ ] No e3nn import anywhere under `packages/` (asserted by test); no import of legacy `mace/` outside `tests/parity/`.
- [ ] The `sphericart` evaluation is written down: fp64 parity result, second-derivative check, the solid-vs-spherical + `[2, 0, 1]` convention mapping develop relies on, CPU timing vs native, and a clear adopt/defer recommendation.

**Inventory gaps assigned here:**
- [ ] `--apply_cutoff` (`mace/tools/arg_parser.py:185`) behaviour is covered (pairs with the P0-7 case) — untested today.
- [ ] `ChebyshevBasisGeneral` — reachable only through the magnetic model today and untested in isolation; it gets its own cases here. Correct its inventory row: it is **not** a `--radial_type` basis (`--radial_type` is `bessel|gaussian|chebyshev`, `arg_parser.py:155-161`).

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/test_radial.py packages/mace-torch/tests/test_embedding.py -v
python -m pytest packages/mace-torch/tests/test_sph_harm.py -v          # parity vs legacy convention, gradcheck/gradgradcheck
python -m pytest tests/parity -m "not slow" -k "radial or sph_harm"     # vs frozen legacy mace/
```
**Review focus:** the spherical-harmonics normalization/ordering conventions (a silent mismatch poisons every downstream parity test in the keystone) and second-derivative correctness; cutoff smoothness at `r_max`; and the recorded `backward_second_derivatives` decision.

---

**Comment by steffen-wedig** [comment]
Some comments so far from starting to work on it for later discussion: 

1. **Block input contract.** The v1 blocks take `node_atomic_numbers` (int64 per node) plus `edge_index`, and every distance argument is `edge_lengths` (`[n_edges, 1]`, Angstrom). Legacy took a one-hot plus a species table and did an argmax in every block. The graph contract in `mace_core` carries the one-hot, so ARCH-2 has to derive the atomic numbers once per batch. Confirm this is the interface ARCH-2 should build on.
2. **Configs for basis, transform and cutoff * The string kinds (`bessel`, `gaussian`, `chebyshev`; `none`, `agnesi`, `soft`) are `Literal`s for now and the `RadialEmbeddingBlock` constructor selects the modules from them. The proper shape is a discriminated union of pydantic models on `kind`, one per basis with only its own parameters, built into modules by factories in the config or somewhere else. 
3. **One Chebyshev class.** Legacy `ChebychevBasis` and `ChebyshevBasisGeneral` compute the same T₁..Tₙ; Shoudl there be a single `ChebyshevBasis(num_basis, include_constant=False)` in v1 to reduce duplication 
4. **Node embedding weight convention.** v1 weight equals the legacy e3nn weight divided by `sqrt(num_elements)`. FM-00's converter must apply this.
5. **sphericart-torch 1.0.9 pinned by magnetic mace*, but probably the magnetic mace model should not use spericart by default, but the reference spherical harmonics backend. 
6. Dropped legacy arguments: `ChebychevBasis.r_max` (never used) and the deprecated `ZBLBasis(r_max=)`. Kinds are lowercase.
7. Modules still read `torch.get_default_dtype()` at construction, as legacy does. That is the seam BKD-2's `PrecisionConfig` replaces.Howt to handle precision settings now?
